### ISSUE TITLE
[Question] Package_ID, PREV, RREV - howto

### ISSUE THREAD

**Issue by iiknd** [conan-io/conan#5368]
Conan version 1.14.1
CONAN_REVISIONS_ENABLED=1

Hi,

either I've not understood the docs correctly or I have missed something, but I have still some gaps to understand Package_ID, PREV, RREV from command line usage perspective.

I've prepared a few packages (local conan server)

`$ conan search mylib/1.0.0@user/channel --revisions -r local_conan_server`
`Revisions for 'mylib/1.0.0@user/channel' at remote 'local_conan_server':
7e6327418b8687e7b54093b22fb2cc9d (2019-06-17 08:51:06 UTC)  // 3
916a7f0ec8349fe381caf04f14e811a7 (2019-06-17 08:27:27 UTC)  // 2
16190d2f99de16dfcb0a2650b22ddc9d (2019-06-17 08:13:02 UTC)  // 1`

//1 first version
//2 recipe update
//3 recipe update and binary package update as well

### Question 1

The documentation says:

> Normally to indicate a binary package of a package you would need something like: <package_name>/<version>@<user_name>/<channel>:<package_id>
> This model has not changed and **you will still be able to use this package reference in the commands** and recipes

`$ conan search mylib/1.0.0@user/channel -r local_conan_server`
`Existing packages for recipe mylib/1.0.0@user/channel:`

`Existing recipe in remote 'local_conan_server':`

`    Package_ID: **68e01ecee23f4aefc3ef9463419259baef83eafb**`

If I now try to install that package as the documentation says:

`conan install -if ~/conan_testinstall mylib/1.0.0@user/channel:68e01ecee23f4aefc3ef9463419259baef83eafb -r local_conan_server`
`ERROR: Conanfile not found at /home/foobar/<my_work_area>/mylib/1.0.0@user/channel:68e01ecee23f4aefc3ef9463419259baef83eafb`

It fails and for some reason it points to my working area?

In case this format for installing conan packages is not supported I think the documentation a bit unclear/misleading?

### Question 2 - "Garbage" collection

Say I have a project with modules: A -> B -> C

I'll search for revisions of A and decide to remove the oldest one (e.g. 16190d2f99de16dfcb0a2650b22ddc9d). Before deletion I wan't to remove also the transitive dependencies (built by me) so I'll check the following:

`conan search A/1.0.0@user/channel#16190d2f99de16dfcb0a2650b22ddc9d -r local_conan_server`
`        [requires]
            B/1.0.0@user/channel:b9e209c73aebfda62c0a0ff6943f62384fc0702e`

I'll check the B next (if it has dependencies I need to remove those as well, recursion):

`conan search B/1.0.0@user/channel#b9e209c73aebfda62c0a0ff6943f62384fc0702e -r local_conan_server`

`ERROR: Not found: http://localhost:9300/v2/conans/B/1.0.0@user/channel/revisions/a3396c3ba0f514de555f3660da8b7ae5301c72f5/files/conanmanifest.txt. [Remote: local_conan_server]`

Obviously I'm mixing PREV, RREV, Package_ID now .... but what would be the way to clean old revisions in a clever way including the dependencies? Do I need to search package revisions for every module and from there find which revision contains the Package_ID?

### Question 3

PREV and RREV, if I search packages with --revisions and get a list, those are  package revisions or recipe revisions, or some combination? Where/how I can figure out which package revision belongs to certain recipe revision if the --revisions gives a flat list of checksums?

**Comment by jgsogo** [comment]
I'll try to have a look to all of these questions, but probably I will end up asking for help because I'm not so experienced with revisions... but I think it makes sense to do it this way so I can be closer to your POV.

### From question 1
Some tips related to your comments:
 * You cannot choose the package_id to install using `conan install <ref>:<pck_id>` as the `pck_id` is computed from the settings, options,... You can do it using the recipe revision, so `conan install <ref>#<rrev>` works.

    :cop: Nevertheless, probably the error message given by `conan install <ref>:<pck_id>` should be improved, I'll open an issue for that.

Question for you:
 * Can you point me to the place where your quote from the documentation is taken:

    > Normally to indicate a binary package of a package you would need something like: <package_name>/@<user_name>/:<package_id>
This model has not changed and you will still be able to use this package reference in the commands and recipes"_

### From question 2

Here you are mixing the `package_id` and the `recipe revision` (as you say). The command `conan search <ref>@<rrev>` returns the packages IDs this reference depends on, so you get it depends on `B/1.0.0@user/channel:b9e209c73aebfda62c0a0ff6943f62384fc0702e` (it is a `<ref>:<pck_id>`). Afterward you are trying to use the `pck_id` as a `rrev` and that is the reason why it cannot find it.

Usually you are to not going to require a `rrev` explicitly (although you can), but it is not usual to have something like:

```python
class Lib(ConanFile):
    name = "A"
    version = "1.0.0"

    requires = "B/1.0.0@user/channel#<recipe revision>"
```

so, there is nothing like _old revisions of dependencies_. Regarding revisions each package is independent of the others, so you can list the requires of your package `A` (it returns `B/1.0.0@user/channel`) and then delete all revisions but the last one of reference `B/1.0.0@user/channel`.

A different question would be, can I find which revisions contain a given package ID? I think this is not possible right now, you should iterate yourself all the revisions and check which ones contain the package ID you are looking for.

### Question 3

The command `conan search <ref> --revisions -r remote` shows the revisions for the recipe (shows _recipe revisions_).

Then the command `conan search <ref>#<rrev> -r remote` shows the packages available for the give recipe revision.

And then the command `conan search <ref>#<rrev>:<pck_id> -r remote --revisions` will list the package revisions for the given package ID of the given recipe revision

----

Let me know if I've been able to solve any doubt/question, so we can keep working on the remaining ones.

Thanks

**Comment by unzap** [comment]
Hi @jgsogo and thanks for the answers!

> Can you point me to the place where your quote from the documentation is taken:

https://blog.conan.io/2019/03/12/New-conan-release-1-13.html

My mistake in my original question, it was in the blog. I found that when I tried to find "revisions" feature information.

I was now able to test the use cases as you kindly answered (Question 3).
One question though:

The blog mentions

> This means that now you can point to an exact revision of a binary package writing the reference and package ID together with the recipe and package revisions: <package_name>/<version>@<user_name>/<channel>#<RREV>:<package_id>#<PREV>

If I run in my env:

`conan search ModuleA/1.0@user/channel#cc5ded7df3dd849678c916fafc05031a:45179f013485e14cb18fc7c13bb09dd7a043acfd -r=my_server --revisions`
`cfbc34296c1333c45ba557822fecc2da (2019-07-01 10:46:06 UTC)`
`c7fc94e279a7dfb5bfd85d456e6aa28f (2019-07-01 09:50:43 UTC)`

So then I could point to:

ModuleA/1.0@user/channel#cc5ded7df3dd849678c916fafc05031a:45179f013485e14cb18fc7c13bb09dd7a043acfd#cfbc34296c1333c45ba557822fecc2da

What is the real usage for this full reference? I found "conan get" being the only command that works with this?

Edit: I think it should allow the use case that CI pushes snapshots to server. Then if consumers are consuming some packages their downloads won't be affected if CI pushes new snapshots simultaneously, as those are completely new files (revisions). Consumers can finish downloading & deploying their current revision of the package(s).

**Comment by unzap** [comment]
About the "garbage collection"/deleting old (unused revisions) still..
An example:

Modules for product 1 (requires chain):
A1 -> B1 -> C1 -> ......... -> X1
Modules for product 2 (requires chain):
A2 -> B2 -> C2 -> ......... -> X2

Add multiple different multi-module products into equation, CI producing snapshots continuously.
Over time you'll have a ton of old revisions of each module in the cache/remote which you want to clean up.

Then you'll check one module from the cache/remote:

A/1.0.0@user/channel
__rrev: 7baa94aca76b803643033750c7500082 - 2019-07-04T09:19:10Z"
____pid: 595b5b6656e9e157701996a12aca465be71677a1"
______prev: 55ff3e072f300be001d329ea9d2bfd9b - 2019-07-04T09:21:40Z
____pid: c3c60f77f47239a8a477ecf3d7ee9bf80212cb8b
______prev: 262cfdac475ab207dafbb7b011512f7c - 2019-07-04T11:03:47Z
____pid: c46e37d71ca0c0aad173b8d196bacb35136b824f
______prev: 9692bbb1021c345f62197cd567397d27 - 2019-07-04T09:40:00Z
__rrev: 64eb55a12feeedf0a600fb79ed4b803b - 2019-07-03T06:01:34Z
____pid: 595b5b6656e9e157701996a12aca465be71677a1
______prev: ea01262829e11373a2d73fd35e8b3510 - 2019-07-03T06:04:01Z
__rrev: 19b71011afd555d58e748dfdf1df1f3c - 2019-07-03T05:47:54Z
____pid: 595b5b6656e9e157701996a12aca465be71677a1
______prev: 6c96a2d8821dc1738bc58417438b1393 - 2019-07-03T05:50:21Z
______prev: 884bbffc9fc49fb546acf498efb8c4b3 - 2019-07-02T05:50:21Z
______prev: 30bda8cde4f083a306fa0d669df8fd60 - 2019-07-01T05:50:21Z

Ok, pick the oldest RREV & PREV (last line above). if you want to delete this PREV you do want to check it's dependencies as well and delete those (revisions).

>conan get A/1.0.0@user/channel#19b71011afd555d58e748dfdf1df1f3c:595b5b6656e9e157701996a12aca465be71677a1#30bda8cde4f083a306fa0d669df8fd60

This gives you:

>[full_requires]
    B/1.0.0@user/channel:c3c60f77f47239a8a477ecf3d7ee9bf80212cb8b

Ok, I can see the Package_ID (c3c60f77f47239a8a477ecf3d7ee9bf80212cb8b), but I can't see which revision of **this** Package_ID it is depending on?

If I'd now run "conan remove B/1.0.0@user/channel:c3c60f77f47239a8a477ecf3d7ee9bf80212cb8b" it would remove all PREVs (package revisions) for that specific Package_ID although I'd want to delete only the one which is "required" by the A/1.0.0@user/channel#19b71011afd555d58e748dfdf1df1f3c:595b5b6656e9e157701996a12aca465be71677a1#30bda8cde4f083a306fa0d669df8fd60?

Or is still something I haven't fully understood? @memsharded , @lasote ? :)

If I'd blindly scan oldest revision(s) from cache/remote it could be that I delete something wrongly if I don't check the dependency chains..

**Comment by lasote** [comment]
My perspective about "cleaning" is:

- The server-side should have its own procedures to clean old artifacts, like the described here: https://jfrog.com/knowledge-base/artifactory-cleanup-methods-how-do-i-delete-old-artifacts/
- The client cache it is only a cache, a cache accumulates stuff but you should be able to just drop it at some point, we have to take it in mind, smart deletion based on the dependency graph would carry a lot of complexity. Nevertheless, there are some ongoing discussions to provide a cleaning mechanism for the local cache: https://github.com/conan-io/conan/issues/5453

**Comment by jgsogo** [comment]
@unzap, now with the [lockfiles](https://docs.conan.io/en/latest/versioning/lockfiles.html) released **as experimental** in `v1.17` you can know which are the revisions used to build a package. These `conan.lock` files that we are generating are intended for Conan usage only, but they contain the full reference with revisions.

**It is experimental**, the structure of the `conan.lock` file may change, or even the whole _lockfile_ implementation can change, but we would love to receive feedback about the feature itself and other usages around it. If it is useful to list all the packages (full references) from a `conan.lock` file, maybe we can consider adding a subcommand to the `conan graph` group or just documenting some bits of the structure of the json file, but we need more feedback related to the core functionality before adding more stuff around it.
